daemon/internal/authconfig/authconfig.go

```diff
@@ -0,0 +1,56 @@
+package authconfig
+
+import (
+	"bytes"
+	"encoding/base64"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+
+	"github.com/moby/moby/api/types/registry"
+	"github.com/moby/moby/v2/errdefs"
+)
+
+// DecodeAuthConfig decodes base64url encoded ([RFC4648, section 5]) JSON
+// authentication information as sent through the X-Registry-Auth header.
+//
+// This function always returns an [AuthConfig], even if an error occurs. It is up
+// to the caller to decide if authentication is required, and if the error can
+// be ignored.
+//
+// [RFC4648, section 5]: https://tools.ietf.org/html/rfc4648#section-5
+func DecodeAuthConfig(authEncoded string) (*registry.AuthConfig, error) {
```

Comment:
I didn't rename this one yet; was on the fence if it perhaps should be called `DecodeHeader` or `DecodeAuthHeader` (or something) to indicate it's specifically for the format used in the `X-Registry-Auth` header.